File: leetcode/322_Coin_Change.py
from typing import List
from unittest import TestCase

class Solution:
    def coinChange(self, coins: List[int], amount: int) -> int:
        coins = sorted([coin for coin in coins if coin <= amount])
        m = len(coins)
        if amount == 0:
            return 0
        elif m == 0:
            return -1
        dp = [-1] * (amount + 1)
        dp[0] = 0
        for i in range(1, amount + 1):
            mod_cnt = 0
            for coin in coins:
                if coin <= i:
                    if dp[i - coin] != -1 and (dp[i] == -1 or dp[i - coin] + 1 < dp[i]):
                        dp[i] = dp[i - coin]  + 1
                else:
                    break
                if i % coin == 0:
                    mod_cnt += 1
            if mod_cnt == m:
                if dp[amount % i] != -1:
                    return dp[amount % i] + int(amount / i) * dp[i]
                else:
                    return -1
        return dp[amount]

class CoinChangeTest(TestCase):
    def test_1(self):
        self.assertEqual(3, Solution().coinChange([1,2,5], 11))

    def test_2(self):
        self.assertEqual(-1, Solution().coinChange([2], 3))

    def test_3(self):
        self.assertEqual(0, Solution().coinChange([1], 0))

    def test_4(self):
        self.assertEqual(2, Solution().coinChange([2], 4))

File: leetcode/322_Coin_Change.rs
use std::collections::VecDeque;
struct Solution {}

impl Solution {
    pub fn coin_change(coins: Vec<i32>, amount: i32) -> i32 {
        if amount == 0 {
            return 0;
        }
        let coins = coins.into_iter()
            .filter_map(|coin| if coin <= amount { Some(coin as usize) } else { None })
            .collect::<Vec<_>>();
        let amount= amount as usize;

        let mut visited: Vec<Option<usize>> = vec![None; amount + 1];
        coins.iter().for_each(|&coin| visited[coin] = Some(1));
        let mut queue = coins.iter().map(|&coin| (1usize, coin)).collect::<VecDeque<_>>();
        while let Some((cnt, v)) = queue.pop_front() {
            if v == amount {
                return cnt as i32;
            }
            for &coin in coins.iter() {
                if coin <=  amount - v {
                    let u =  v + coin;
                    if visited[u].is_none() {
                        visited[u] = Some(cnt + 1);
                        queue.push_back((cnt + 1, u));
                    }
                } 
            }
        }
        -1
    }

    pub fn _coin_change_1(coins: Vec<i32>, amount: i32) -> i32 {
        if amount == 0 {
            return 0;
        }
        let coins = coins.into_iter()
            .filter_map(|coin| if coin <= amount { Some(coin as usize) } else { None })
            .collect::<Vec<_>>();
        let amount= amount as usize;

        let mut dp: Vec<Option<usize>> = vec![None; amount + 1];
        for &coin in coins.iter() {
            dp[coin] = Some(1);
        }
        for i in 1..=amount {
            if let Some(i_cnt) = dp[i] {
                for &coin in coins.iter().filter(|&&coin| coin <= amount - i) {
                    let next_amount = i + coin;
                    dp[next_amount] = dp[next_amount].map_or(Some(i_cnt + 1), |next_cnt| Some(next_cnt.min(i_cnt + 1)));
                }
            }
        }
        dp[amount].map_or(-1, |cnt| cnt as i32)
    }
}

#[test]
fn test_1() {
    assert_eq!(3, Solution::coin_change(vec![1,2,5], 11));
}

#[test]
fn test_2() {
    assert_eq!(-1, Solution::coin_change(vec![2], 3));
}
#[test]
fn test_3() {
    assert_eq!(0, Solution::coin_change(vec![1], 0));
}

#[test]
fn test_4() {
    assert_eq!(2, Solution::coin_change(vec![1,2147483647], 2));
}
